libparsec/crates/client/src/certif/user_revoke.rs

```diff
@@ -0,0 +1,170 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+use libparsec_client_connection::ConnectionError;
+use libparsec_protocol::authenticated_cmds;
+use libparsec_types::prelude::*;
+
+use super::{
+    store::CertifStoreError, CertifOps, CertificateBasedActionOutcome, InvalidCertificateError,
+    InvalidKeysBundleError,
+};
+use crate::EventTooMuchDriftWithServerClock;
+
+#[derive(Debug, thiserror::Error)]
+pub enum CertifRevokeUserError {
+    #[error("Component has stopped")]
+    Stopped,
+    #[error("Cannot reach the server")]
+    Offline,
+    #[error("Cannot revoke oneself")]
+    UserIsSelf,
+    #[error("User not found")]
+    UserNotFound,
+    #[error("Author not allowed")]
+    AuthorNotAllowed,
+    #[error("Our clock ({client_timestamp}) and the server's one ({server_timestamp}) are too far apart")]
+    TimestampOutOfBallpark {
+        server_timestamp: DateTime,
+        client_timestamp: DateTime,
+        ballpark_client_early_offset: f64,
+        ballpark_client_late_offset: f64,
+    },
+    #[error("There is no key available in this realm for encryption")]
+    NoKey,
+    #[error(transparent)]
+    InvalidKeysBundle(#[from] Box<InvalidKeysBundleError>),
+    #[error(transparent)]
+    InvalidCertificate(#[from] Box<InvalidCertificateError>),
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+}
+
+impl From<ConnectionError> for CertifRevokeUserError {
+    fn from(value: ConnectionError) -> Self {
+        match value {
+            ConnectionError::NoResponse(_) => Self::Offline,
+            // TODO: handle organization expired and user revoked here ?
+            err => Self::Internal(err.into()),
+        }
+    }
+}
+
+impl From<CertifStoreError> for CertifRevokeUserError {
+    fn from(value: CertifStoreError) -> Self {
+        match value {
+            CertifStoreError::Stopped => Self::Stopped,
+            CertifStoreError::Internal(err) => err.into(),
+        }
+    }
+}
+
+pub(super) async fn revoke_user(
+    ops: &CertifOps,
+    user: UserID,
+) -> Result<CertificateBasedActionOutcome, CertifRevokeUserError> {
+    if *ops.device.device_id.user_id() == user {
+        return Err(CertifRevokeUserError::UserIsSelf);
+    }
+
+    // Loop is needed to deal with server requiring greater timestamp
```

Comment:
Maybe we can use a generic function which handle greater timestamp
fn foo<T>(...) -> T {
    loop { ... }
}

Reply:
I'd rather keep it simple given how fast Rust + async becomes out of hand

Typically your `foo` method has to be asynchronous and capture the lifetime of the closure's parameter, I had to do that for certificate store and it involve unsafe stuff (in this case they it was worth it because it bring guarantee on how data should be locked, while here it's just a convenient helper)

https://github.com/Scille/parsec-cloud/blob/ea898fa300fc2292b78a13d81450fe7ad34b1d5f/libparsec/crates/client/src/certif/store.rs#L158-L229